kono-api/src/api/v1/image/image.control.js

```diff
@@ -81,4 +88,49 @@ export const count = async (req, res) => {
         res.send({ msg: 'server error' });
     }
 
-}
\ No newline at end of file
+}
+
+/* For post upload. */
+
+class MimetypeError extends Error {
+    constructor(...params) {
+        super(...params);
+        this.message = 'Image file mimetype undefined or unsupported.';
+    }
+}
+
+const storage = multer.diskStorage({
+    destination: (req, file, cb) => {
+        cb(null, process.env.UPLOAD_DIR)
+    },
+    filename: (req, file, cb) => {
+        if (!UPLOAD_MIME_TYPES.includes(file.mimetype)) {
+            return cb(new MimetypeError(), '');
+        }
+        const extension = file.mimetype.split('/').pop();
+        cb(null, `${uniqueString()}.${extension}`);
+    }
+});
+
+const uploadFiles = multer({ 
+    storage: storage,
+    limits: {
+        fileSize: UPLOAD_MAX_FILE_SIZE
+    }
+}).array(UPLOAD_KEY, UPLOAD_MAX_NUM_FILES);
+
+export const upload = async (req, res) => {
+    uploadFiles(req, res, (err) => {
+        if (err) {
+            res.status(400);
```

Comment:
Assuming that it is highly likely that an error occurred during the middleware uploadFiles might be caused by the client's fault, I set the status to be 400 in this case. But I'm not sure if this is okay since there could be some scenarios that error occurs by the server's fault (e.g. Out of disk space). Could there be better way to set error status is a more finely grained way?

Reply:
HTTP Status Code에서 error handling 할 때 크게 4xx (client error)와 5xx (server error)로 나뉘어지는데
지금 그걸 구분할 수 있는 로직이 없는 것 같아요.

예를 들어 위에 `storage` 함수만 보면
- 이미지 파일이 최대 용량 (10MB)을 넘지 않고 올바른 확장자(.jpg, .png)를 가지고 있는지 (근데 JPEG도 지원해야되는 거 아닌가? JPG랑 JPEG랑 똑같은데)
    - 아마 용량은 `multer`에서 핸들링 하는 것 같은데...
    - 그래도 error message를 구분해야 할 필요성은 있어 보임
    - 400 Bad Request
- File system error (서버 공간 부족, disk failure, nonexistent directory, duplicate file name 등)
    - `multer` 내부에서 돌아가다가 error 뱉을 것 같은데 `err.toString()`으로 때우는 건 좀 그래요
    - 500 Internal Server Error

그리고 찾아보니 서버에서 파일 형식이 올바르고 해석할 수 있었지만 용량이 너무 큰 경우에는 `422 Unprocessible Entity`라는 status code가 있다고 하네요

Reply:
고쳐야 할것들이 많군요 ㅋㅋㅋ 나머지는 간단히 수정할 것 같은데 여기가 좀 고쳐야 할 게 많네요 우선 파일 사이즈는 multer limits 으로 걸어놨는데
음... field, file size, file 개수를 explicit 하게 검사해서 400 처리 나머지를 500 으로 하는건 어떤가요?
마임타입 image/jpg 은 추가할게요

Reply:
그럼 illegal한 경우가 아예 `multer` 로직 내부로 안 들어가게 한다는 건가요

Reply:
아니요 multer 안에서 특정 Error (저 `MimeTypeError` 를 client fault error 포괄하는 좀더 제너럴한 에러러 확장 후) 발생시키고 그 에러인 경우 400 나머지는 500 이런식으로요
multer limits 등으로 하지않고 수동으로 체크해서 error 를 발생시키는거죠

Reply:
https://github.com/expressjs/multer/blob/master/lib/multer-error.js
`MulterError`가 client fault 에만 발생하는 에러라
`err instanceof multer.MulterError` 로 체크하고 + 추가하고 싶은 클라이언트 에러(e.g. 파일익스텐션)를 위해 extend class 만들고 (이거 두개는 400),
그 이외의 경우를 500 으로 처리하겠습니다.

Reply:
프론트에서 `err.toString()`에 따라 다른 텍스트를 띄워야 할 것 같아요
대표적인 경우에만 어떤 메시지 뜨는지 doc에 기록해주세요 b